#16 - login/register form

PiotrFedak opened the pull request:
User login and registration

This should close #16.

EwelinaLasowy commented on resources/css/style.css:
```diff
@@ -0,0 +1,151 @@
+.container-of-login {
```
Why you don't use TailwindCSS?

EwelinaLasowy commented on app/Http/Controllers/RegisterController.php:
```diff
@@ -0,0 +1,25 @@
+ 
+}
+
+?>
```
```suggestion

```

EwelinaLasowy commented on routes/api.php:
```diff
@@ -5,5 +5,11 @@
+
+Route::post('/register', [RegisterController::class, 'SignUp']);
+Route::post('/login', [LoginController::class,'Login']);
+Route::post('/logout', [LoginController::class, 'Logout']);
```
```suggestion
Route::post('/logout', [LoginController::class, 'Logout']);

```

EwelinaLasowy commented on app/Http/Controllers/LoginController.php:
```diff
@@ -0,0 +1,33 @@
+
+}
+
+?>
```
```suggestion

```

EwelinaLasowy commented on app/Http/Controllers/LoginController.php:
```diff
@@ -0,0 +1,33 @@
+        $credentials = $request -> validate ([
+            'email' => ['required', 'email'],
+            'password' => ['required'], ]);
+        //Authentication
```
```suggestion
```

I don't think we need this comment here. This comment add nothing.

EwelinaLasowy commented on app/Http/Controllers/LoginController.php:
```diff
@@ -0,0 +1,33 @@
+    {
+        $credentials = $request -> validate ([
+            'email' => ['required', 'email'],
+            'password' => ['required'], ]);
```
Codestyle? Please run `composer csf`.

EwelinaLasowy commented on public/index.php:
```diff
@@ -4,6 +4,7 @@
 
 use Illuminate\Contracts\Http\Kernel;
 use Illuminate\Http\Request;
+use Illuminate\support\Facades\Redis;
```
```suggestion
```

EwelinaLasowy commented on resources/js/Pages/Login.vue:
```diff
@@ -0,0 +1,98 @@
+      this.showLoginForm = !this.showLoginForm;
+    },
+    loginUser() {
+      axios
```
We don't need axios, because we have Inertia.js in the project. Please refactor this.

EwelinaLasowy reviewed (commented):
1. Maybe if buttons will be aligned, it will be okay. But now I don't like it.

![image](https://github.com/blumilksoftware/escooters/assets/56546832/775de386-33af-4cef-8046-43cf4ad35bf5)

2. We can use TailwindUI to style components. It will be easier and speed up a work.

3. When I fill too long string to inputs and click "Register" button I see error "Failed."

![image](https://github.com/blumilksoftware/escooters/assets/56546832/6bca792c-60ad-4bdb-b7a6-ba37fc2869ca)
 
 But why? We should provide some feedback to user, what is wrong.
 
 And to be honest, I'm not a big fan of this confirmation dialog - can we change it?

kamilpiech97 commented on app/Http/Controllers/LoginController.php:
```diff
@@ -0,0 +1,33 @@
+{
+    public function Login(Request $request)
+    {
+        $credentials = $request -> validate ([
+            'email' => ['required', 'email'],
+            'password' => ['required'], ]);
```
I would move that into `FormRequest` - https://laravel.com/docs/10.x/validation#form-request-validation

PiotrFedak replied:
done

kamilpiech97 commented on app/Http/Controllers/LoginController.php:
```diff
@@ -0,0 +1,33 @@
+        Auth::logout();
+
+        return response()->json(['status' => true, 'message' => "Logged out"]);
+
```
Why here u have space between return and end of the function but above in function not.
Please run `composer csf`

EwelinaLasowy commented:
Also please merge main to this pr - I think that now GitHub Actions should work like a charm.

kamilpiech97 commented on routes/api.php:
```diff
@@ -5,5 +5,11 @@
 
 Route::middleware("auth:sanctum")->get("/user", fn(Request $request): JsonResponse => new JsonResponse($request->user()));
+
+Route::post('/register', [RegisterController::class, 'SignUp']);
+Route::post('/login', [LoginController::class,'Login']);
+Route::post('/logout', [LoginController::class, 'Logout']);
```
We are creating API or inertia app?

kamilpiech97 replied:
Take a look at how it works in the toby [app](https://github.com/blumilksoftware/toby)

kamilpiech97 commented on app/Http/Controllers/RegisterController.php:
```diff
@@ -0,0 +1,25 @@
+use Hash;
+class RegisterController extends Controller
+{
+    public function SignUp(Request $request)
```
```suggestion
    public function signUp(Request $request)
```
name of the function should be written in `camelCase`

kamilpiech97 replied:
this function would be better named as a `register`

EwelinaLasowy commented:
Please fix errors from PHP linter and JS linter (you can use commands `composer csf` and `npm run lintf`

EwelinaLasowy commented:
While committing the suggestion form code review, you can add suggestions to batch and make just one commit with all changes - please read about it [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request).

PiotrFedak commented:
frontend will be changed soon to match with the rest of the site 

- added Inertia instead of API 
- small fixes like codestyle
- added new errors messages
![Screenshot_20230615_125255](https://github.com/blumilksoftware/escooters/assets/115737114/5a00382a-5396-4455-838c-b7e24d917d0c)

kamilpiech97 commented:
conflicts @PiotrFedak

kamilpiech97 commented on app/Http/Controllers/Auth/LogoutController.php:
```diff
@@ -0,0 +1,25 @@
+        $request->session()->invalidate();
+        $request->session()->regenerateToken();
+
+        Session::flash("success", "Logout Complite.");
+
+        return Inertia::location("/");
```
```suggestion
        return redirect()->route("login");
```
That should be enough.

kamilpiech97 commented on app/Http/Requests/LoginRequest.php:
```diff
@@ -0,0 +1,31 @@
+
+class LoginRequest extends FormRequest
+{
+    /**
+     * Determine if the user is authorized to make this request.
+     */
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array|string>
+     */
```
```suggestion
```

kamilpiech97 commented on app/Http/Requests/RegisterRequest.php:
```diff
@@ -0,0 +1,33 @@
+
+class RegisterRequest extends FormRequest
+{
+    /**
+     * Determine if the user is authorized to make this request.
+     */
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array|string>
+     */
```
```suggestion
```

kamilpiech97 commented on resources/js/Pages/Auth/Login.vue:
```diff
@@ -0,0 +1,76 @@
+  password: '',
+  errors: {},
+})
+
+const alertMessage = ref('')
+
+const submit = async () => {
+  try {
+    const response = await form.post('/login', {
+      onError: () => {
+        alertMessage.value = 'Error'
+      },
+      onSuccess: () => {
+        console.log('You have successfully logged in')
+        window.location.href = '/'
+      },
+    })
+    console.log(response)
+  } catch (error) {
+    console.log('An error occured', error)
+  }
+}
+
+watch(() => form.errors, (errors) => {
+  console.log(errors)
+})
+</script>
+
+<template>
+  <div class="bg-gradient-to-br from-blue-600 to-indigo-600">
+    <head title="Log In" />
+
+    <main class="custom-main grid min-h-screen place-items-center">
+      <div class="float-right mt-4 w-1/2">
+        <section class="mx-auto max-w-md rounded-xl border bg-white p-4">
+          <h1 class="mb-4 text-2xl">
+            Log In
+          </h1>
+
+          <form @submit.prevent="submit">
```
```suggestion
})

function attemptLogin() {
  form.post('/login')
}
</script>

<template>
  <div class="bg-gradient-to-br from-blue-600 to-indigo-600">
    <head title="Log In" />

    <main class="custom-main grid min-h-screen place-items-center">
      <div class="float-right mt-4 w-1/2">
        <section class="mx-auto max-w-md rounded-xl border bg-white p-4">
          <h1 class="mb-4 text-2xl">
            Log In
          </h1>

          <form @submit.prevent="attemptLogin">
```

kamilpiech97 replied:
That should be enough.
You don't need to initialize errors in the form.

kamilpiech97 commented on resources/js/Pages/Auth/Signup.vue:
```diff
@@ -0,0 +1,62 @@
+import { useForm } from '@inertiajs/vue3'
+import { watch } from 'vue'
+
+const form = useForm({
+  name: '',
+  email: '',
+  password: '',
+  errors: {},
+})
+
+const submit = async () => {
+  try { 
+    form.post('/register') 
+  } catch (error) {
+    console.log(error)
+  }
+}
+
+watch(() => form.errors, (errors) => {
+  console.log(errors)
+})
+
+</script>
+
+<template>
+  <div class="flex h-screen w-full items-center justify-center bg-gradient-to-br from-blue-600 to-indigo-600">
+    <form @submit.prevent="submit">
```
Same as above with form, errors etc.

kamilpiech97 commented on resources/js/Pages/Auth/UserStatus.vue:
```diff
@@ -0,0 +1,25 @@
+<script>
+export default {
+  computed: {
+    isAuthenticated() {
+      return this.$page.props.auth.loggedIn
+    },
+    user() {
```
```suggestion
<script>
export default {
  computed: {
    isAuthenticated() {
      return this.$page.props.auth.loggedIn
    },
    user() {
```
I would create page for logged users only, login page should be for not logged users.

kamilpiech97 commented on resources/views/app.blade.php:
```diff
@@ -7,6 +7,6 @@
+    @inertia
 </body>
-</html>
+</html>
```
```suggestion
</html>

```

kamilpiech97 commented on routes/web.php:
```diff
@@ -2,7 +2,20 @@
 
-Route::get("/", fn(): Response => inertia("Welcome"));
+Route::get("/signup", fn(): Response => inertia::render("Auth/Signup"));
+Route::get("/login", [LoginController::class, "create"])->name("login");
+Route::post("/login", [LoginController::class, "login"]);
+
+Route::post("/register", [RegisterController::class, "store"])->name("register");
+
+Route::get("/", fn() => Inertia::render("Welcome"))->name("home");
+
+Route::get("/login", [LoginController::class, "create"])->name("login");
+Route::post("/logout", [LogoutController::class, "logout"])->name("logout");
```
Duplicated?

PiotrFedak commented on app/Http/Controllers/Auth/LoginController.php:
```diff
@@ -0,0 +1,34 @@
+        return Inertia::render("Auth/Login");
+    }
+
+    public function login(LoginRequest $request)
```
Is like this, okey ?

`public function login(LoginRequest $request): RedirectResponse`
`return redirect()->route("home");`

EwelinaLasowy commented on tests/Feature/LoginTest.php:
```diff
@@ -0,0 +1,40 @@
+        $response->assertStatus(200);
+    }
+
+    public function testOfLogin(): void
```
And where we creating a user? I don't see any ` $user = User::factory()->create();` or something like that. We should also check, if after login user is really logged into app. I mean we can use assertion `->assertAuthenticated()`.

PiotrFedak replied:
Okey, my bad. I changed to User::factory like this.

```
 public  function testUserCanLoginWithValidCredentials(): void
    {
        $user = User::factory()->create([
            "email" => "email@example.com",
            "password" => bcrypt("password@example"),
        ]);

        $response = $this->post("/login", [
            "email" => "email@example.com",
            "password" => "password@example",
        ]);

        $response->assertStatus(302);
        $this->assertAuthenticatedAs($user);
    }
```

EwelinaLasowy commented on tests/Feature/LoginTest.php:
```diff
@@ -0,0 +1,40 @@
+    public function testOfLogin(): void
+    {
+        $response = $this->post("/login", [
+            "email" => "Email@example.com",
```
```suggestion
            "email" => "email@example.com",
```

EwelinaLasowy commented on tests/Feature/SignupTest.php:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * A basic feature test example.
+     */
+    public function testOfSignupDisplay(): void
```
```suggestion
    public function testShowSignupPage(): void
```

EwelinaLasowy commented on tests/Feature/SignupTest.php:
```diff
@@ -0,0 +1,30 @@
+        $response->assertStatus(200);
+    }
+
+    public function testSuccesfullSignup(): void
```
```suggestion
    public function testUserCanSignupWithProperCredentials(): void
```

EwelinaLasowy reviewed (commented):
Don't be afraid of long test names. Test names should describe what the test does, without reading the entire test body.
Please add some test to unhappy path - like login with wrong credentials,  signup with for example too long name, authenticated user should be redirected form login page to dashboard. We can also test that user after signup is logged in the app, guest can't see dashboard page and so on.

PiotrFedak commented on app/Http/Middleware/HandleInertiaRequests.php:
```diff
@@ -35,6 +35,10 @@ public function version(Request $request): ?string
     {
         return array_merge(parent::share($request), [
+            "auth" => [
+                "loggedIn" => auth()->check(),
```
okay i added some routes to middleware, but that variable  will stay couse i use that to display at the top right corner who is already logged in.

EwelinaLasowy commented on tests/Feature/SignupTest.php:
```diff
@@ -0,0 +1,66 @@
+    public function testUserCannotBeCreatedWithInvalidName(): Void
+    {
+        $response = $this->post("/register", [
+            "name" => "phasellus faucibus scelerisque eleifend donec pretium vulputate sapien nec sagittis aliquam malesuada bibendum arcu vitae elementum curabitur vitae nunc sed velit dignissim sodales ut eu sem integer vitae justo eget magna fermentum iaculis eu non diam phasellus vestibulum lorem sed risus ultricies tristique nulla aliquet enim tortor at auctor urna nunc id cursus metus aliqu ...
```
Instead of writing this long string, you can user for example Str::random(256) - it will generate string 256 characters long.

EwelinaLasowy commented on tests/Feature/SignupTest.php:
```diff
@@ -0,0 +1,66 @@
+            "password" => bcrypt("password@example"),
+        ]);
+
+        $response->assertStatus(302);
```
You can also use `assertSessionHasErrors("someKey"); ` assertion to check what errors has the session.

EwelinaLasowy commented on tests/Feature/SignupTest.php:
```diff
@@ -0,0 +1,66 @@
+    {
+        $user = User::factory()->create();
+
+        $this->actingAs($user);
+
+        $response = $this->get("/dashboard");
+
+        $response->assertStatus(200);
+        $this->assertAuthenticated();
```
```suggestion
        $this->actingAs($user)
                 ->get("/dashboard");
                 ->assertStatus(200);
                 ->assertAuthenticated();
```

EwelinaLasowy replied:
Maybe spaces are not okay, but I think tests in this form will be cleaner.

kamilpiech97 commented on composer.json:
```diff
@@ -73,4 +73,4 @@
     "minimum-stability": "stable",
     "prefer-stable": true
-}
+}
```
```suggestion
}

```

kamilpiech97 commented on resources/js/Pages/Auth/Login.vue:
```diff
@@ -0,0 +1,78 @@
+
+function attemptLogin() {
+  form.post('/login', {
+    onError: function(err) {
+      error.value=err
+    },
```
`form.errors.{field}` should work

kamilpiech97 commented on resources/js/Pages/Auth/Signup.vue:
```diff
@@ -0,0 +1,69 @@
+    console.log(error)
+  }
+}
+watch(() => form.errors, (errors) => {
+  console.log(errors)
+})
```
Do we need this `watch` and `console.log()` here and above?

PiotrFedak replied:
okay, i get rid of it

EwelinaLasowy commented on resources/js/Pages/Auth/Signup.vue:
```diff
@@ -0,0 +1,67 @@
+            </h1>
+            <button type="button" class="mr-2 inline-flex items-center rounded-lg bg-blue-700 p-2.5 text-center text-sm font-medium text-white hover:bg-blue-800 focus:outline-none focus:ring-4 focus:ring-blue-300 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800">
+              <a href="/">back </a>
+              <svg aria-hidden="true" class="h-5 w-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
```
Instead of using svg, we can use icon here. I'd recommend heroicons. But the design will change as you said, so it's just a piece of advise for the future.

EwelinaLasowy commented on resources/js/Pages/Welcome.vue:
```diff
@@ -1,15 +1,70 @@
+          this.$inertia.visit('/')
+        })
+        .catch((error) => {
+          console.error(error)
```
:eyes:

EwelinaLasowy reviewed (commented):
Please remember, that author comment should resolve the conversation.

kamilpiech97 commented on resources/js/Pages/Auth/Signup.vue:
```diff
@@ -0,0 +1,67 @@
+
+<script setup>
+import { useForm } from '@inertiajs/vue3'
+
+const form = useForm({
```
Lets name that view `Register.vue` instead of `Singup.vue`.

MichalMyskow commented on resources/js/Pages/Auth/UserStatus.vue:
```diff
@@ -0,0 +1,24 @@
+<script>
+export default {
+  computed: {
+    isAuthenticated() {
+      return this.$page.props.auth.isAuth
+    },
+    user() {
+      return this.$page.props.auth.user
+    },
+  },
+}
+</script>
```
For this example, you can do this as follows:

```
<script setup>
import { usePage } from '@inertiajs/vue3'
import { computed } from 'vue'

const isAuthenticated = computed(() => usePage().props.auth.isAuth)
const user = computed(() => usePage().props.auth.user)
</script>
```
